File: java-context-os/src/main/java/com/owencli/contextos/memory/FactMemory.java
package com.owencli.contextos.memory;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.owencli.contextos.core.model.FactRecord;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

/**
 * Fact Memory — KV store with versioning for structured user facts.
 * <p>
 * Unlike append-only memory, Fact Memory preserves the timeline
 * so the agent always reads the current_value, not the history.
 * <p>
 * Storage: SQLite table "facts" with columns:
 *   id, type, current_value, history(JSON), confidence, status, source, created_at, updated_at
 */
public class FactMemory {

    private static final Logger log = LoggerFactory.getLogger(FactMemory.class);
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final SQLiteStore store;
    private final String userId;

    public FactMemory(SQLiteStore store, String userId) {
        this.store = store;
        this.userId = userId;
        ensureTable();
        log.info("FactMemory initialized");
    }

    private void ensureTable() {
        // Facts are stored in the same memories table with type="fact"
        // This reuses the existing SQLiteStore infrastructure
    }

    /**
     * Set a fact. If it already exists (same type), update it with versioning.
     * If it doesn't exist, create a new entry.
     */
    public CompletableFuture<FactRecord> setFact(String type, String value, double confidence, String source) {
        return getFact(type).thenCompose(existing -> {
            if (existing.isPresent()) {
                // Update existing fact with versioning
                FactRecord record = existing.get();
                record.update(value, confidence, source);
                return saveFactRecord(record).thenApply(v -> record);
            } else {
                // Create new fact
                FactRecord record = new FactRecord(type, value);
                record.setConfidence(confidence);
                record.setSource(source);
                return saveFactRecord(record).thenApply(v -> record);
            }
        });
    }

    /**
     * Get the current value of a fact by type.
     */
    public CompletableFuture<Optional<FactRecord>> getFact(String type) {
        return store.queryMemories("fact", null, userId, null, null, null, 10, 0)
                .thenApply(results -> results.stream()
                        .map(this::rowToFactRecord)
                        .filter(Objects::nonNull)
                        .filter(r -> r.getType().equals(type))
                        .filter(FactRecord::isActive)
                        .max(Comparator.comparing(FactRecord::getUpdatedAt)));
    }

    /**
     * Get all active facts.
     */
    public CompletableFuture<List<FactRecord>> getAllFacts() {
        return store.queryMemories("fact", null, userId, null, null, null, 200, 0)
                .thenApply(results -> results.stream()
                        .map(this::rowToFactRecord)
                        .filter(Objects::nonNull)
                        .filter(FactRecord::isActive)
                        .sorted(Comparator.comparing(FactRecord::getType))
                        .collect(Collectors.toList()));
    }

    /**
     * Get fact history for a given type.
     */
    public CompletableFuture<List<String>> getFactHistory(String type) {
        return getFact(type).thenApply(opt ->
                opt.map(FactRecord::getHistory).orElse(List.of()));
    }

    /**
     * Retrieve facts relevant to a query (for context building).
     * <p>
     * Only facts whose type or value matches the query keywords are returned.
     * The unconditional "user.*" catch-all has been removed — facts must now
     * pass a keyword relevance check to avoid injecting irrelevant personal
     * data (e.g. user.name=张三) into unrelated tasks like "写 K8s Deployment".
     */
    public CompletableFuture<List<FactRecord>> retrieve(String query, int topK) {
        String q = query.toLowerCase();
        return getAllFacts().thenApply(facts -> {
            var scored = facts.stream()
                    .filter(f -> f.getType().toLowerCase().contains(q)
                            || f.getCurrentValue().toLowerCase().contains(q))
                    .map(f -> {
                        // Compute keyword overlap score for ranking
                        double score = computeKeywordOverlap(q, f);
                        // Use the higher of keyword overlap or stored confidence
                        // as the effective relevance signal
                        return new AbstractMap.SimpleEntry<>(f, score);
                    })
                    .sorted(Map.Entry.<FactRecord, Double>comparingByValue().reversed())
                    .limit(topK)
                    .map(Map.Entry::getKey)
                    .collect(Collectors.toList());
            log.debug("FactMemory.retrieve: query='{}', matched={}/{}", q, scored.size(), facts.size());
            return scored;
        });
    }

    /**
     * Compute keyword overlap between the query and a fact's type + value.
     * Returns a value in [0, 1] where 1.0 means the query fully matches.
     */
    private double computeKeywordOverlap(String query, FactRecord fact) {
        if (query == null || query.isBlank()) return 0.0;
        String text = (fact.getType() + " " + fact.getCurrentValue()).toLowerCase();
        String[] queryTokens = query.trim().split("\\s+");
        if (queryTokens.length == 0) return 0.0;
        long matchCount = java.util.Arrays.stream(queryTokens)
                .filter(token -> token.length() >= 2 && text.contains(token))
                .count();
        // Base confidence + keyword bonus, capped at 1.0
        return Math.min(1.0, fact.getConfidence() * 0.5 + (double) matchCount / queryTokens.length * 0.5);
    }

    /**
     * Save a fact record to the store.
     */
    private CompletableFuture<Void> saveFactRecord(FactRecord record) {
        try {
            String memId = record.getId();
            Map<String, Object> metadata = new LinkedHashMap<>();
            metadata.put("fact_type", record.getType());
            metadata.put("current_value", record.getCurrentValue());
            metadata.put("history", record.getHistory());
            metadata.put("confidence", record.getConfidence());
            metadata.put("fact_status", record.getStatus());
            metadata.put("source", record.getSource());
            metadata.put("created_at", record.getCreatedAt().toString());
            metadata.put("updated_at", record.getUpdatedAt().toString());

            String content = String.format("[Fact] %s = %s (confidence=%.2f, status=%s)",
                    record.getType(), record.getCurrentValue(), record.getConfidence(), record.getStatus());

            return store.saveMemory(memId, "fact", content, null, userId, null, metadata, null)
                    .thenAccept(id -> log.debug("Fact saved: {} = {}", record.getType(), record.getCurrentValue()));
        } catch (Exception e) {
            log.warn("Failed to save fact record: {}", e.getMessage());
            return CompletableFuture.completedFuture(null);
        }
    }

    private FactRecord rowToFactRecord(Map<String, Object> row) {
        try {
            var meta = parseMetadata(row.get("metadata"));
            if (meta == null || meta.isEmpty()) return null;

            var record = new FactRecord();
            record.setId((String) row.get("id"));
            record.setType((String) meta.getOrDefault("fact_type", "unknown"));
            record.setCurrentValue((String) meta.getOrDefault("current_value", ""));
            record.setHistory(parseStringList(meta.get("history")));
            record.setConfidence(toDouble(meta.getOrDefault("confidence", 0.8)));
            record.setStatus((String) meta.getOrDefault("fact_status", "ACTIVE"));
            record.setSource((String) meta.getOrDefault("source", ""));
            try {
                record.setCreatedAt(Instant.parse((String) meta.getOrDefault("created_at", Instant.now().toString())));
                record.setUpdatedAt(Instant.parse((String) meta.getOrDefault("updated_at", Instant.now().toString())));
            } catch (Exception e) {
                record.setCreatedAt(Instant.now());
                record.setUpdatedAt(Instant.now());
            }
            return record;
        } catch (Exception e) {
            log.warn("Failed to parse fact record: {}", e.getMessage());
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    private List<String> parseStringList(Object obj) {
        if (obj instanceof List<?> list) {
            return list.stream().map(Object::toString).collect(Collectors.toList());
        }
        return new ArrayList<>();
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseMetadata(Object metadata) {
        if (metadata instanceof Map m) return (Map<String, Object>) m;
        if (metadata instanceof String s) {
            try { return MAPPER.readValue(s, new TypeReference<>() {}); }
            catch (Exception e) { return Map.of(); }
        }
        return Map.of();
    }

    private double toDouble(Object obj) {
        if (obj instanceof Number n) return n.doubleValue();
        try { return Double.parseDouble(obj.toString()); }
        catch (Exception e) { return 0.0; }
    }

    /**
     * Get summarized output for all facts as context.
     */
    public CompletableFuture<String> getFactsSummary() {
        return getAllFacts().thenApply(facts -> {
            if (facts.isEmpty()) return "";
            var sb = new StringBuilder("Known facts about user:\n");
            for (var f : facts) {
                sb.append("- ").append(f.getType()).append(": ")
                        .append(f.getCurrentValue())
                        .append(" (confidence=").append(String.format("%.2f", f.getConfidence()))
                        .append(")\n");
            }
            return sb.toString();
        });
    }
}

File: context_os/memory/fact_memory.py
"""FactMemory — 版本化事实 KV 存储。

表由 SQLiteStore._DDL 统一创建。
"""

from __future__ import annotations

import json
import uuid
from datetime import datetime, timezone
from typing import Optional

from context_os.core.logger import get_logger
from context_os.memory.store import SQLiteStore

logger = get_logger(__name__)


class FactRecord:
    """版本化事实记录。"""

    def __init__(
        self,
        id: str = "",
        content: str = "",
        category: str = "",
        confidence: float = 1.0,
        version: int = 1,
        user_id: str = "anonymous",
        source: str = "",
        metadata: Optional[dict] = None,
        created_at: Optional[str] = None,
    ):
        self.id = id
        self.content = content
        self.category = category
        self.confidence = confidence
        self.version = version
        self.user_id = user_id
        self.source = source
        self.metadata = metadata or {}
        self.created_at = created_at or datetime.now(timezone.utc).isoformat()

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "content": self.content,
            "category": self.category,
            "confidence": self.confidence,
            "version": self.version,
            "user_id": self.user_id,
            "source": self.source,
            "metadata": self.metadata,
            "created_at": self.created_at,
        }


class FactMemory:
    """版本化事实 KV 存储。

    Args:
        store: SQLite 存储层实例。
        user_id: 默认用户 ID。
    """

    def __init__(self, store: SQLiteStore, user_id: str = "anonymous"):
        self.store = store
        self.user_id = user_id
        logger.info("FactMemory initialized (user=%s)", user_id)

    @staticmethod
    def _row_to_record(row: dict) -> FactRecord:
        """将查询行转换为 FactRecord。"""
        return FactRecord(
            id=row["id"],
            content=row.get("current_value") or row.get("content", ""),
            category=row["category"],
            confidence=row["confidence"],
            version=row["version"],
            user_id=row["user_id"],
            source=row.get("source", ""),
            metadata=json.loads(row.get("metadata", "{}")),
            created_at=row["created_at"],
        )

    async def set(
        self,
        fact_id: str,
        content: str,
        category: str,
        confidence: float = 1.0,
        source: str = "",
        metadata: Optional[dict] = None,
    ) -> FactRecord:
        """设置/更新一条事实（版本递增）。"""
        now = datetime.now(timezone.utc).isoformat()
        existing = await self.store.query(
            "SELECT * FROM facts WHERE id = ?", [fact_id]
        )

        if existing:
            row = existing[0]
            history = json.loads(row.get("history", "[]"))
            history.append({
                "value": row.get("current_value") or row.get("content", ""),
                "version": row["version"],
                "updated_at": now,
            })
            new_version = row["version"] + 1
            await self.store.execute(
                "UPDATE facts SET content = ?, current_value = ?, version = ?, confidence = ?, "
                "metadata = ?, history = ?, updated_at = ? WHERE id = ?",
                [
                    content, content, new_version, confidence,
                    json.dumps(metadata or {}), json.dumps(history), now, fact_id,
                ],
            )
            logger.debug("Fact updated: id=%s, v%d -> v%d", fact_id, row["version"], new_version)
            return FactRecord(
                id=fact_id, content=content, category=category,
                confidence=confidence, version=new_version,
                user_id=self.user_id, source=source, metadata=metadata,
            )

        # 新事实
        await self.store.execute(
            "INSERT INTO facts (id, content, category, confidence, version, "
            "user_id, source, metadata, current_value, history, created_at, updated_at) "
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            [
                fact_id, content, category, confidence, 1, self.user_id,
                source, json.dumps(metadata or {}), content, json.dumps([]), now, now,
            ],
        )
        logger.info("Fact created: id=%s, category=%s", fact_id, category)
        return FactRecord(
            id=fact_id, content=content, category=category,
            confidence=confidence, version=1,
            user_id=self.user_id, source=source,
            metadata=metadata, created_at=now,
        )

    async def get(self, fact_id: str) -> Optional[FactRecord]:
        """按 ID 获取事实。"""
        rows = await self.store.query(
            "SELECT * FROM facts WHERE id = ?", [fact_id]
        )
        return self._row_to_record(rows[0]) if rows else None

    async def query(self, category: Optional[str] = None, limit: int = 100) -> list[FactRecord]:
        """按类别查询事实。"""
        if category:
            rows = await self.store.query(
                "SELECT * FROM facts WHERE user_id = ? AND category = ? "
                "ORDER BY updated_at DESC LIMIT ?",
                [self.user_id, category, limit],
            )
        else:
            rows = await self.store.query(
                "SELECT * FROM facts WHERE user_id = ? "
                "ORDER BY updated_at DESC LIMIT ?",
                [self.user_id, limit],
            )
        return [self._row_to_record(r) for r in rows]

    async def delete(self, fact_id: str) -> None:
        """删除一条事实。"""
        await self.store.execute(
            "DELETE FROM facts WHERE id = ?", [fact_id]
        )
        logger.debug("Fact deleted: id=%s", fact_id)
